compat/test/browser/suspense.test.js

```diff
@@ -290,6 +293,104 @@ describe('suspense', () => {
 		});
 	});
 
+	it('should properly call lifecycle methods and maintain state of a delayed suspending component', () => {
+		let componentWillMount = sinon.spy();
+		let componentDidMount = sinon.spy();
+		let componentDidUpdate = sinon.spy();
+		let componentWillUnmount = sinon.spy();
+
+		/** @type {() => void} */
+		let increment;
+
+		/** @type {() => Promise<void>} */
+		let resolve;
+		let resolved = false;
+		const promise = new Promise(_resolve => {
+			resolve = () => {
+				resolved = true;
+				_resolve();
+				return promise;
+			};
+		});
+
+		class LifecycleSuspender extends Component {
+			constructor(props) {
+				super(props);
+				this.state = { count: 0 };
+
+				increment = () => this.setState(({ count }) => ({ count: count + 1 }));
+			}
+			render() {
+				if (this.state.count == 2 && !resolved) {
+					throw promise;
+				}
+
+				return (
+					<Fragment>
+						<p>Count: {this.state.count}</p>
+					</Fragment>
+				);
+			}
+			componentWillMount() {
+				componentWillMount();
+			}
+			componentDidMount() {
+				componentDidMount();
+			}
+			componentWillUnmount() {
+				componentWillUnmount();
+			}
+			componentDidUpdate() {
+				componentDidUpdate();
+			}
+		}
+
+		render(
+			<Suspense fallback={<div>Suspended...</div>}>
+				<LifecycleSuspender />
+			</Suspense>,
+			scratch
+		);
+
+		expect(scratch.innerHTML).to.eql(`<p>Count: 0</p>`);
+		expect(componentWillMount).to.have.been.calledOnce;
+		expect(componentDidMount).to.have.been.calledOnce;
+		expect(componentDidUpdate).to.not.have.been.called;
+		expect(componentWillUnmount).to.not.have.been.called;
+
+		increment();
+		rerender();
+
+		expect(scratch.innerHTML).to.eql(`<p>Count: 1</p>`);
+		expect(componentWillMount).to.have.been.calledOnce;
+		expect(componentDidMount).to.have.been.calledOnce;
+		expect(componentDidUpdate).to.have.been.calledOnce;
+		expect(componentWillUnmount).to.not.have.been.called;
+
+		increment();
+		rerender();
+
+		expect(scratch.innerHTML).to.eql(`<div>Suspended...</div>`);
+		expect(componentWillMount).to.have.been.calledOnce;
+		expect(componentDidMount).to.have.been.calledOnce;
+		expect(componentDidUpdate).to.have.been.calledOnce;
+		expect(componentWillUnmount).to.not.have.been.called;
+
+		return resolve().then(() => {
+			rerender();
+
+			expect(scratch.innerHTML).to.eql(`<p>Count: 2</p>`);
+			expect(componentWillMount).to.have.been.calledOnce;
+			expect(componentDidMount).to.have.been.calledOnce;
+			// TODO: This is called thrice since the cDU queued up after the second
+			// increment is never cleared once the component suspends. So when it
+			// resumes and the component is rerendered, we queue up another cDU so
+			// cDU is called an extra time.
```

Comment:
Note there is a follow up to this comment in #2098 (specifically [this comment](https://github.com/preactjs/preact/pull/2098/files#r345516634))